### PR TITLE
Partial aggregation support

### PR DESCRIPTION
**Pull request by AlexanderSaydakov**
No upgrade scripts yet. I will submit that as a separate pull request.

### CODE REVIEW

**Comment by AlexanderSaydakov** [comment]
I prepared the update script from 1.5.0 to 1.6.0 and added to this pull request since it is still open.

File: sql/datasketches--1.5.0--1.6.0.sql [hunk 1]
```diff
@@ -0,0 +1,1350 @@
...
+
+CREATE OR REPLACE FUNCTION cpc_sketch_build_agg(internal, anyelement) RETURNS internal
+    AS '$libdir/datasketches', 'pg_cpc_sketch_build_agg'
+    LANGUAGE C IMMUTABLE PARALLEL SAFE;
+
+CREATE OR REPLACE FUNCTION cpc_sketch_build_agg(internal, anyelement, int) RETURNS internal
+    AS '$libdir/datasketches', 'pg_cpc_sketch_build_agg'
+    LANGUAGE C IMMUTABLE PARALLEL SAFE;
+
+CREATE OR REPLACE FUNCTION cpc_sketch_from_internal(internal) RETURNS cpc_sketch
```
Comment (jmalkin) [inline sql/datasketches--1.5.0--1.6.0.sql:34]: based on kll, i expected a change to serialize?

Comment (jmalkin) [reply sql/datasketches--1.5.0--1.6.0.sql:34]: Nevermind, it's pretty clearly something around state management for distinct counting in particular. I'm sure it'll make more sense when I review the specific code around the implementation.

File: src/aod_sketch_pg_functions.c [hunk 2]
```diff
@@ -118,89 +125,43 @@ Datum pg_aod_sketch_add_item(PG_FUNCTION_ARGS) {
...
-    lg_k = PG_NARGS() > 3 ? PG_GETARG_INT32(3) : 0;
+    stateptr = palloc(sizeof(struct aod_agg_state));
+    stateptr->type = MUTABLE_SKETCH;
+    stateptr->lg_k = PG_NARGS() > 3 ? PG_GETARG_INT32(3) : 0;
+    stateptr->num_values = arr_len;
     p = PG_NARGS() > 4 ? PG_GETARG_FLOAT4(4) : 1;
-    if (lg_k) {
-      sketchptr = p ? aod_sketch_new_lgk_p(arr_len, lg_k, p) : aod_sketch_new_lgk(arr_len, lg_k);
+    if (stateptr->lg_k) {
+      stateptr->ptr = p ? aod_sketch_new_lgk_p(arr_len, stateptr->lg_k, p) : aod_sketch_new_lgk(arr_len, stateptr->lg_k);
```
Comment (jmalkin) [inline src/aod_sketch_pg_functions.c:134]: I think this will always take the first branch unless p is explicitly set to 0?

Comment (AlexanderSaydakov) [reply src/aod_sketch_pg_functions.c:134]: yeah, I think this is a mistake, but harmless. I will fix

File: src/cpc_sketch_pg_functions.c [hunk 3]
```diff
@@ -286,7 +368,6 @@ Datum pg_cpc_sketch_union(PG_FUNCTION_ARGS) {
     cpc_sketch_delete(sketchptr2);
   }
   sketchptr = cpc_union_get_result(unionptr);
-  cpc_union_delete(unionptr);
```
Comment (jmalkin) [inline src/cpc_sketch_pg_functions.c:289]: Unclear why this was removed? It'll be cleaned up at the end of the transaction, but it seems to have been a temporary local variable?

Comment (AlexanderSaydakov) [reply src/cpc_sketch_pg_functions.c:289]: now get_result deletes the union

File: src/hll_sketch_pg_functions.c [hunk 4]
```diff
@@ -242,43 +209,105 @@ Datum pg_hll_sketch_get_estimate_from_internal(PG_FUNCTION_ARGS) {
...
 
-  stateptr = (struct hll_union_state*) PG_GETARG_POINTER(0);
+  stateptr1 = (struct hll_agg_state*) PG_GETARG_POINTER(0);
+  stateptr2 = (struct hll_agg_state*) PG_GETARG_POINTER(1);
+
+  stateptr = palloc(sizeof(struct hll_agg_state));
+  stateptr->type = SKETCH;
+  stateptr->lg_k = stateptr1 ? stateptr1->lg_k : stateptr2->lg_k;
+  stateptr->tgt_type = stateptr1 ? stateptr1->tgt_type : stateptr2->tgt_type;
+  stateptr->ptr = hll_union_new(stateptr->lg_k);
```
Comment (jmalkin) [inline src/hll_sketch_pg_functions.c:248]: A comment on why we're creating a union after setting `stateptr->type = SKETCH;` might be useful, even though reading on shows that we overwrite it with the union result.

File: src/theta_sketch_pg_functions.c [hunk 5]
```diff
@@ -287,47 +343,98 @@ Datum pg_theta_union_get_result(PG_FUNCTION_ARGS) {
...
   oldcontext = MemoryContextSwitchTo(aggcontext);
 
-  unionptr = PG_GETARG_POINTER(0);
-  sketchptr = theta_union_get_result(unionptr);
-  bytes_out = theta_sketch_serialize(sketchptr, VARHDRSZ);
-  theta_sketch_delete(sketchptr);
-  theta_union_delete(unionptr);
+  stateptr = (struct agg_state*) PG_GETARG_POINTER(0);
+  if (stateptr->type == MUTABLE_SKETCH) {
+    stateptr->ptr = theta_sketch_compact(stateptr->ptr);
```
Comment (jmalkin) [inline src/theta_sketch_pg_functions.c:352]: "how is a compact sketch mutable?" -- had to look at `theta_sketch_compact()` separately but that answered it

**Review by jmalkin — APPROVED** [review]
That was a lot to try reading though in one shot.

Most of it is fairly straightforward (once I remember PG is C rather than C++). There were a few things that could benefit from an explicit comment on why we're doing things. But definitely not a blocker for merging.

**Pull request merged**